src/huggingface_hub/repocard.py

```diff
@@ -1,30 +1,288 @@
+        """
+
+        body = {
+            "repoType": repo_type,
+            "content": str(self),
+        }
+        headers = {"Accept": "text/plain"}
+
+        try:
+            r = requests.post(
+                "https://huggingface.co/api/validate-yaml", body, headers=headers
+            )
+            r.raise_for_status()
+        except requests.exceptions.HTTPError as exc:
+            if r.status_code == 400:
+                raise RuntimeError(r.text)
+            else:
+                raise exc
+
+    def push_to_hub(
+        self,
+        repo_id,
+        token=None,
+        repo_type="model",
+        commit_message=None,
+        commit_description=None,
+        revision=None,
+        create_pr=None,
+    ):
+        """Push a RepoCard to a Hugging Face Hub repo.
+
+        Args:
+            repo_id (`str`):
+                The repo ID of the Hugging Face Hub repo to push to. Example: "nateraw/food".
+            token (`str`, *optional*):
+                Authentication token, obtained with `huggingface_hub.HfApi.login` method. Will default to
+                the stored token.
+            repo_type (`str`, *optional*):
+                The type of Hugging Face repo to push to. Defaults to None, which will use
+                use "model". Other options are "dataset" and "space".
+            commit_message (`str`, *optional*):
+                The summary / title / first line of the generated commit
+            commit_description (`str`, *optional*)
+                The description of the generated commit
+            revision (`str`, *optional*):
+                The git revision to commit from. Defaults to the head of the
+                `"main"` branch.
+            create_pr (`bool`, *optional*):
+                Whether or not to create a Pull Request with this commit. Defaults to `False`.
+        Returns:
+            `str`: URL of the commit which updated the card metadata.
+        """
+
+        if repo_type is None:
+            repo_type = "model"
+
+        if repo_type not in ["model", "space", "dataset"]:
+            raise RuntimeError(
+                "Provided repo_type '{repo_type}' should be one of ['model', 'space',"
+                " 'dataset']."
+            )
+
+        # Validate card before pushing to hub
+        self.validate(repo_type=repo_type)
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            tmp_path = Path(tmpdir) / REPOCARD_NAME
+            tmp_path.write_text(str(self))
+            url = upload_file(
+                path_or_fileobj=str(tmp_path),
+                path_in_repo=REPOCARD_NAME,
+                repo_id=repo_id,
+                token=token,
+                repo_type=repo_type,
+                commit_message=commit_message,
+                commit_description=commit_description,
+                create_pr=create_pr,
+                revision=revision,
+            )
+        return url
+
+    @classmethod
+    def from_template(
```

Comment:
No strong opinion to be honest. `from_data` makes sense to me. WDYT @nateraw ?

Reply:
kinda like `from_data` too

Reply:
I'm not a big fan of `from_data`, as I think it's a bit more misleading. We aren't really creating the card 100% from data - a lot of the content is going to come from the template. And keeping the name with "template" in it at least makes sense when you get an error about not having Jinja2 (imo). 

I could be convinced to change it, though, if anybody is very opinionated about it!

Reply:
Seems no one is so opinionated but we still don't have a consensus though 😁. 
Maybe `RepoCard.render_with_data` (or `RepoCard.render_from_data(...)`) is more convincing ? It's less the name that I would give to a classmethod but we have both the templating and the data aspects.

Reply:
And in any case, if no naming reach a perfect consensus I suggest with keep `RepoCard.from_template` just because it's good enough and so that we can close this (huge) PR :)

Reply:
As an alternative design, the pattern we chose in `skops` is for the user to create an instance of the object, then add contents for the sections through instance methods. From the users' perspective, the code looks like this, and the implementation is here: https://github.com/skops-dev/skops/blob/main/skops/card/_model_card.py

``` python
# this constructor would take a path to a template in a future iteration, for now supporting only
# the default template.
# Instead of calling the metadata `CardData`, we call it `metadata` which is easier
# to understand from the users' perspective.
model_card = card.Card(model=sklearn_estimator, metadata=...)

# The user can also access and change/add metadata by accessing the metadata object
model_card.metadata.license = "mit"

limitations = "This model is not ready to be used in production."
model_description = (
    "This is a HistGradientBoostingClassifier model trained on breast cancer dataset."
    " It's trained with Halving Grid Search Cross Validation, with parameter grids on"
    " max_leaf_nodes and max_depth."
)
model_card_authors = "skops_user"
get_started_code = (
    "import pickle\nwith open(pkl_filename, 'rb') as file:\n    clf = pickle.load(file)"
)
citation_bibtex = "bibtex\n@inproceedings{...,year={2020}}"

# they can add info in a single call, or chain calls to `add`, since all `add*` methods
# return self.
model_card.add(
    citation_bibtex=citation_bibtex,
    get_started_code=get_started_code,
    model_card_authors=model_card_authors,
    limitations=limitations,
    model_description=model_description,
)

# adding metrics
eval_descr = (
    "The model is evaluated on test data using accuracy and F1-score with macro"
    " average."
)
model_card.add(eval_method=eval_descr)

...
model_card.add_metrics(**{"accuracy": accuracy, "f1 score": f1})

# or save and add plots
...
model_card.add_plot(**{"Confusion matrix": "confusion_matrix.png"})

# or add tabular information about hyper parameter search space and resuts
...
model_card.add_table(
    # they decide if the table should be in a `<details>` tag.
    folded=True,
    **{
        "Hyperparameter search results": cv_results,
        "Classification report": clf_report,
    },
)

# And finally, this is how the model is _rendered_ into a file.
model_card.save(Path(local_repo) / "README.md")

# However, they could also render the model into a string:
rendered = model_card.render()
```

Reply:
I prefer that we leave it as `from_template` for now - I am opinionated enough to push for that. I think it makes it very clear that we're using a 'template', which implies the need for the extra `Jinja2` dep. I promise I'm not just being lazy and not wanting to change the name 😅 .

As for the usage described by @adrinjalali - I prefer the way we have implemented here, as it results in less code I think, and effectively does the same thing.